async_fuse/src/memfs/cache/mod.rs

```diff
@@ -0,0 +1,742 @@
+        .overflow_add(time.tv_nsec.overflow_div(1_000_000))
+}
+
+/// A memory block collection
+struct MemBlockBucket {
+    /// The inner is read-write lock protected vector of `MemBlock`
+    inner: Arc<RwLock<Vec<Option<MemBlock>>>>,
+}
+
+impl Hash for MemBlockBucket {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (*self.inner).data_ptr().hash(state);
+    }
+}
+
+impl PartialEq<MemBlockBucket> for MemBlockBucket {
+    fn eq(&self, other: &Self) -> bool {
+        (*self.inner).data_ptr() == (*other.inner).data_ptr()
+    }
+}
+
+impl Eq for MemBlockBucket {}
+
+impl Clone for MemBlockBucket {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::<
+                parking_lot::lock_api::RwLock<parking_lot::RawRwLock, Vec<Option<MemBlock>>>,
+            >::clone(&self.inner),
+        }
+    }
+}
+
+/// A vector holding `MEMORY_BUCKET_VEC_SIZE` `MemBlock`, wrapped in an Arc.
+/// Arc is used to shared this vector with priority queue.
+impl MemBlockBucket {
+    #[allow(dead_code)]
+    /// Init with None values
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(
+                std::iter::repeat_with(|| None)
+                    .take(MEMORY_BUCKET_VEC_SIZE)
+                    .collect(),
+            )),
+        }
+    }
+
+    /// Insert one `MemBlock` into the `index` position of the bucket
+    #[allow(dead_code)]
+    pub(crate) fn insert(&mut self, index: usize, mem: MemBlock) {
+        self.inner.write().insert(index, Some(mem));
+    }
+}
+
+impl Deref for MemBlockBucket {
+    type Target = RwLock<Vec<Option<MemBlock>>>;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+/// A wrapper for `IoMemBlock`, which is used for I/O operations
+pub struct IoMemBlock {
+    /// The inner `MemBlock` that contains data
+    inner: Option<MemBlock>,
+    /// The offset for this `MemBlock`
+    offset: usize,
+    /// The end offset for this `MemBlock`
+    end: usize,
+}
+
+impl Debug for IoMemBlock {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("IoMemBlock")
+            .field("offset", &self.offset)
+            .field("end", &self.end)
+            .finish()
+    }
+}
+
+impl IoMemBlock {
+    /// The constructor of `IoMemBlock`
+    const fn new(inner: Option<MemBlock>, offset: usize, end: usize) -> Self {
+        Self { inner, offset, end }
+    }
+
+    /// Turn `IoMemBlock` into slice, only used in test
+    #[allow(dead_code)]
+    pub(crate) fn as_slice(&self) -> &[u8] {
+        if self.inner.is_none() {
+            panic!("IoMemBlock inner is None, cannot as_slice");
+        }
+
+        if let Some(ref block) = self.inner {
+            let ptr = block.as_ptr_from_offset(self.offset);
+            unsafe { std::slice::from_raw_parts(ptr, self.end.overflow_sub(self.offset)) }
+        } else {
+            panic!("the inner Option<MemBlock> should not be none");
+        }
+    }
+}
```

Comment:
**Internal unsoundness**: No matter where it is used, this method should be `unsafe` because it bypasses the `RwLock`.

For this case, `with`-style methods are better than `unsafe` methods.
https://doc.rust-lang.org/nightly/std/thread/struct.LocalKey.html#method.with

Reply:
It's a good point, we should expose unsafe attribute to the caller.